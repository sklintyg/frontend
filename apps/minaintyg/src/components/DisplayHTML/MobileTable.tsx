/* eslint-disable react/no-array-index-key */
import type { Element, HTMLReactParserOptions } from 'html-react-parser'
import { domToReact } from 'html-react-parser'
import { isElement } from './utils/isElement'

export function MobileTable({ header, body, options }: { header?: Element; body?: Element; options: HTMLReactParserOptions }) {
  if (!header || !body) {
    return null
  }

  const headerRow = header.children.filter(isElement)[0]

  return (
    <>
      {body.children.filter(isElement).map(({ children }, tIndex) => {
        const bodyCells = children.filter(isElement).filter((node) => node.name !== 'th')
        const rowHeader = children.filter(isElement).find((node) => node.name === 'th')
        const headerRows = headerRow?.children.filter(isElement).slice(rowHeader != null ? 1 : 0)

        return (
          <table className="ids-table" key={tIndex}>
            {rowHeader && (
              <thead>
                <tr>
                  <th colSpan={2} className="text-center">
                    {domToReact(rowHeader.children, options)}
                  </th>
                </tr>
              </thead>
            )}
            <tbody>
              {headerRows?.map((node, rIndex) => {
                const cellElement = bodyCells[rIndex]

                if (!cellElement) {
                  return null
                }

                return (
                  <tr key={rIndex}>
                    <th>{domToReact(node.children, options)}</th>
                    <td>{domToReact(cellElement.children, options)}</td>
                  </tr>
                )
              })}
            </tbody>
          </table>
        )
      })}
    </>
  )
}
