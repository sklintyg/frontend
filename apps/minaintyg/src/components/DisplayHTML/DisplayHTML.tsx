/* eslint-disable react/no-array-index-key */
import { IDSIconExternal, IDSLink } from '@frontend/ids-react-ts'
import type { DOMNode } from 'html-react-parser'
import parse, { attributesToProps, domToReact } from 'html-react-parser'
import { createElement } from 'react'
import { MobileTable } from './MobileTable'
import { isElement } from './utils/isElement'

const options = {
  replace: (domNode: DOMNode) => {
    if (isElement(domNode)) {
      const { name, attribs, children } = domNode
      const props = attributesToProps(attribs)
      const match = name.match(/^h(\d)$/)

      if (match) {
        const [tag, level] = match
        return createElement(tag, { className: `ids-heading-${parseInt(level, 10) < 4 ? level : 4}` }, domToReact(children, options))
      }

      if (name === 'table') {
        const elements = children.filter(isElement)
        return (
          <>
            <div className="md:hidden">
              <MobileTable
                header={elements.find((node) => node.name === 'thead')}
                body={elements.find((node) => node.name === 'tbody')}
                options={options}
              />
            </div>
            <div className="hidden md:block">
              <table className="ids-table">{domToReact(children, options)}</table>
            </div>
          </>
        )
      }

      if (name === 'a') {
        return (
          <IDSLink underlined>
            <a {...props}>{domToReact(children, options)}</a>
            {attribs.target === '_blank' && <IDSIconExternal slot="append-icon" size="s" />}
          </IDSLink>
        )
      }

      if (name === 'th' && children.length === 0) {
        return <td />
      }
    }
    return undefined
  },
}

export function DisplayHTML({ html }: { html: string }) {
  return <>{parse(html, options)}</>
}
