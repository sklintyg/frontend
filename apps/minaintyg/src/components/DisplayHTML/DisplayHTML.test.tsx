import { render, screen } from '@testing-library/react'
import { DisplayHTML } from './DisplayHTML'

describe('tables', () => {
  it('Should add class to tables', () => {
    render(<DisplayHTML html="<table></table>" />)
    expect(screen.getByRole('table')).toHaveClass('ids-table', { exact: true })
  })

  it('Should render mobile table', () => {
    const { container } = render(
      <DisplayHTML
        html="`
          <table>
            <thead>
              <tr>
                <th>First cell</th>
                <th>Second cell</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <td>123</td>
                <td>Lorem ipsum</td>
              </tr>
              <tr>
                <td>456</td>
                <td>dolor sit amet</td>
              </tr>
            </tbody>
          </table>
        `"
        mobile
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('Should render mobile table with row headers', () => {
    const { container } = render(
      <DisplayHTML
        html="`
          <table>
            <thead>
              <tr>
                <th></th>
                <th>First cell</th>
                <th>Second cell</th>
                <th>Third cell</th>
              </tr>
            </thead>
            <tbody>
              <tr>
                <th>Row heading</th>
                <td>1</td>
                <td>2</td>
                <td>3</td>
              </tr>
            </tbody>
          </table>
        `"
        mobile
      />
    )
    expect(container).toMatchSnapshot()
  })
})

describe('link', () => {
  it('Should render link as ids-link', () => {
    const { container } = render(<DisplayHTML html="<a href='http://some-place.com'>Some link</a>" />)
    expect(container).toMatchSnapshot()
  })

  it('Should render link with external icon when target is _blank', () => {
    const { container } = render(<DisplayHTML html="<a href='http://some-place.com' target='_blank'>Some link</a>" />)
    expect(container).toMatchSnapshot()
  })
})

describe('headings', () => {
  it.each(Array.from({ length: 6 }, (_, index) => index + 1))('Should render %s heading as expected', (level) => {
    const tag = `h${level}`
    render(<DisplayHTML html={`<${tag}>text</${tag}>`} />)
    expect(screen.getByRole('heading')).toHaveClass(`ids-heading-${level}`, { exact: true })
  })
})
