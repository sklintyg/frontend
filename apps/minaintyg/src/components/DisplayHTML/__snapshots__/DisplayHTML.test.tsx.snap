// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`link > Should render link as ids-link 1`] = `
<div>
  <ids-link>
    <a
      href="http://some-place.com"
    >
      Some link
    </a>
  </ids-link>
</div>
`;

exports[`link > Should render link with external icon when target is _blank 1`] = `
<div>
  <ids-link>
    <a
      href="http://some-place.com"
      target="_blank"
    >
      Some link
    </a>
    <ids-icon-external
      class="ids-icon"
      slot="append-icon"
    />
  </ids-link>
</div>
`;

exports[`tables > Should render mobile table 1`] = `
<div>
  \`
          
  <table
    class="ids-table"
  >
    <tbody>
      <tr>
        <th>
          First cell
        </th>
        <td>
          123
        </td>
      </tr>
      <tr>
        <th>
          Second cell
        </th>
        <td>
          Lorem ipsum
        </td>
      </tr>
    </tbody>
  </table>
  <table
    class="ids-table"
  >
    <tbody>
      <tr>
        <th>
          First cell
        </th>
        <td>
          456
        </td>
      </tr>
      <tr>
        <th>
          Second cell
        </th>
        <td>
          dolor sit amet
        </td>
      </tr>
    </tbody>
  </table>
  
        \`
</div>
`;

exports[`tables > Should render mobile table with row headers 1`] = `
<div>
  \`
          
  <table
    class="ids-table"
  >
    <thead>
      <tr>
        <th
          class="text-center"
          colspan="2"
        >
          Row heading
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <th>
          First cell
        </th>
        <td>
          1
        </td>
      </tr>
      <tr>
        <th>
          Second cell
        </th>
        <td>
          2
        </td>
      </tr>
      <tr>
        <th>
          Third cell
        </th>
        <td>
          3
        </td>
      </tr>
    </tbody>
  </table>
  
        \`
</div>
`;
